File: app/src/main/java/io/digibyte/presenter/activities/AssetImageActivity.java
package io.digibyte.presenter.activities;

import android.animation.ObjectAnimator;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityOptionsCompat;
import androidx.databinding.DataBindingUtil;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.load.resource.bitmap.DownsampleStrategy;
import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.DrawableImageViewTarget;
import com.bumptech.glide.request.target.ImageViewTarget;
import com.bumptech.glide.request.target.Target;

import io.digibyte.R;
import io.digibyte.databinding.ActivityAssetImageBinding;
import io.digibyte.databinding.ActivityQrCodeBinding;
import io.digibyte.presenter.activities.base.BRActivity;
import io.digibyte.presenter.activities.callbacks.ActivityAssetImageCallback;
import io.digibyte.presenter.activities.callbacks.ActivityQRCodeCallback;
import io.digibyte.tools.animation.BRAnimator;
import io.digibyte.tools.qrcode.QRUtils;

public class AssetImageActivity extends BRActivity {

    private static final String ASSET_IMAGE_URL = "AssetImageActivity:AssetImageUrl";
    private ActivityAssetImageBinding binding;

    private ActivityAssetImageCallback callback = this::remove;

    public static void show(AppCompatActivity activity, View view, String assetImageUrl) {
        Intent intent = new Intent(activity, AssetImageActivity.class);
        intent.putExtra(ASSET_IMAGE_URL, assetImageUrl);
        ActivityOptionsCompat options = ActivityOptionsCompat.
                makeSceneTransitionAnimation(activity, view, "asset_image");
        //Shared object behaves strangely with the gif unfortunately
        activity.startActivity(intent/*, options.toBundle()*/);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        supportPostponeEnterTransition();
        binding = DataBindingUtil.setContentView(this, R.layout.activity_asset_image);
        binding.setCallback(callback);
        String assetImageUrl = getIntent().getStringExtra(ASSET_IMAGE_URL);
        Glide.with(this).load(assetImageUrl).diskCacheStrategy(DiskCacheStrategy.ALL).listener(new RequestListener<Drawable>() {
            @Override
            public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                finish();
                return false;
            }

            @Override
            public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                supportStartPostponedEnterTransition();
                new Handler().postDelayed(() -> {
                    ObjectAnimator colorFade = BRAnimator.animateBackgroundDim(binding.background, false, null);
                    colorFade.setStartDelay(350);
                    colorFade.setDuration(500);
                    colorFade.start();
                }, 250);
                return false;
            }
        }).into(binding.image);
    }

    @Override
    public void onBackPressed() {
        remove();
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.enter_from_left, R.anim.exit_to_left);
    }

    private void remove() {
        BRAnimator.animateBackgroundDim(binding.background, true, this::finish).start();
    }
}

File: app/src/main/java/io/digibyte/presenter/activities/AssetImageActivity.kt
package io.digibyte.presenter.activities

import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import io.digibyte.R
import io.digibyte.databinding.ActivityAssetImageBinding
import io.digibyte.databinding.ActivityAssetVideoPlayerBinding
import io.digibyte.presenter.activities.base.BRActivity
import io.digibyte.presenter.activities.callbacks.ActivityAssetImageCallback
import io.digibyte.presenter.activities.utils.createPlayer
import io.digibyte.presenter.activities.utils.initialize
import io.digibyte.tools.animation.BRAnimator
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import java.util.concurrent.TimeUnit

class AssetImageActivity : BRActivity() {
    private lateinit var background: ViewGroup

    private var videoBinding: ActivityAssetVideoPlayerBinding? = null

    private val callback = ActivityAssetImageCallback { this.remove() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        when (intent.getStringExtra(MIME)) {
            "image" -> {
                val binding: ActivityAssetImageBinding = DataBindingUtil.setContentView(this, R.layout.activity_asset_image)
                background = binding.background
                binding.callback = callback
                val contentUrl = intent.getStringExtra(CONTENT_URL)
                Glide.with(this).load(contentUrl).diskCacheStrategy(DiskCacheStrategy.ALL).listener(object : RequestListener<Drawable> {
                    override fun onLoadFailed(e: GlideException?, model: Any, target: Target<Drawable>, isFirstResource: Boolean): Boolean {
                        finish()
                        return false
                    }

                    override fun onResourceReady(resource: Drawable, model: Any, target: Target<Drawable>, dataSource: DataSource, isFirstResource: Boolean): Boolean {
                        fadeInBackground()
                        return false
                    }
                }).into(binding.image)
            }
            "video" -> {
                fadeInBackground()
                val contentUrl = intent.getStringExtra(CONTENT_URL)
                videoBinding = DataBindingUtil.setContentView(this, R.layout.activity_asset_video_player)
                background = videoBinding!!.background
                videoBinding!!.callback = callback
                videoBinding!!.progress.visibility = View.VISIBLE
                videoBinding!!.videoView.player = createPlayer(this).initialize(
                        context = this,
                        url = contentUrl.replace("ecom", "com"),
                        onPlaybackStarted = {
                            videoBinding?.progress?.visibility = View.GONE
                        },
                        finishPlayback = {
                            remove()
                        },
                        errorPlayback = {
                            Toast.makeText(this, R.string.error_playing_asset_video, Toast.LENGTH_SHORT).show()
                            remove()
                        }
                )
            }
        }
    }

    private fun fadeInBackground() {
        Completable.fromRunnable {
            val colorFade = BRAnimator.animateBackgroundDim(background, false, null)
            colorFade.startDelay = 350
            colorFade.duration = 500
            colorFade.start()
        }.subscribeOn(AndroidSchedulers.mainThread()).delaySubscription(250, TimeUnit.MILLISECONDS).subscribe()
    }

    override fun onBackPressed() {
        remove()
    }

    override fun finish() {
        super.finish()
        videoBinding?.videoView?.player?.release()
        overridePendingTransition(R.anim.enter_from_left, R.anim.exit_to_left)
    }

    private fun remove() {
        BRAnimator.animateBackgroundDim(background, true) { this.finish() }.start()
    }

    companion object {

        private const val CONTENT_URL = "AssetImageActivity:AssetContentUrl"
        private const val MIME = "AssetImageActivity:Mime"

        fun show(activity: AppCompatActivity, view: View, assetContentUrl: String, mime: String) {
            val intent = Intent(activity, AssetImageActivity::class.java)
            intent.putExtra(CONTENT_URL, assetContentUrl)
            intent.putExtra(MIME, mime)
            activity.startActivity(intent)
        }
    }
}
